File: hadoop/src/main/java/systematicchaos/studentmobility/domain/Student.java
/**
 * Student Mobility
 * 
 * Javier Fernández-Bravo Peñuela
 * Universitat Politècnica de València
 * 
 * hadoop - domain/Student.java
 */

package systematicchaos.studentmobility.domain;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeFormatterBuilder;
import java.time.format.DateTimeParseException;
import java.util.Locale;

public class Student {
	
    private String homeInstitution;
    private String countryOfHomeInstitution;
    private Integer age;
    private Gender gender;
    private StudyLevel studyLevel;
    private MobilityType mobilityType;
    private String hostInstitution;
    
    private String countryOfHostInstitution;
    private String countryOfWorkPlacement;
    private Float lengthStudyPeriod;
    private Float lengthWorkPlacement;
    private LocalDate studyStartDate;
    private LocalDate placementStartDate;
    private Boolean taughtHostLang;
    private String languageTaught;
    private Float studyGrant;
    private Float placementGrant;
    
    private static final String EMPTY_STRING = "";
    private static final String DEFAULT_STR = EMPTY_STRING;
    private static final int ZERO = 0;
    private static final int DEFAULT_INT = ZERO;
    private static final boolean DEFAULT_BOOL = Boolean.FALSE;
    private static final String SPLIT = String.valueOf(';');
    
    private static final int NUM_ATTRIBUTES_BASIC = 8;
    private static final int NUM_ATTRIBUTES_FULL = 17;
    
    public Student() {
    }

    /**
     * Constructor with parameters just for compulsory attributes
     */
    public Student(String homeInstitution, String countryOfHomeInstitution,
            Integer age, Gender gender, StudyLevel studyLevel, MobilityType mobilityType,
            String hostInstitution, String countryOfHostInstitution) {
    	this.setHomeInstitution(homeInstitution);
    	this.setCountryOfHomeInstitution(countryOfHomeInstitution);
    	this.setAge(age);
    	this.setGender(gender);
    	this.setStudyLevel(studyLevel);
    	this.setMobilityType(mobilityType);
    	this.setHostInstitution(hostInstitution);
    	this.setCountryOfHostInstitution(countryOfHostInstitution);
    }

    /**
     * Constructor with parameters for all attributes
     */
    public Student(String homeInstitution, String countryOfHomeInstitution,
            Integer age, Gender gender, StudyLevel studyLevel, MobilityType mobilityType,
            String hostInstitution, String countryOfHostInstitution, String countryOfWorkPlacement,
            Float lengthStudyPeriod, Float lengthWorkPlacement,
            LocalDate studyStartDate, LocalDate placementStartDate,
            Boolean taughtHostLang, String languageTaught,
            Float studyGrant, Float placementGrant) {
        this(homeInstitution, countryOfHomeInstitution, age, gender, studyLevel,
        		mobilityType, hostInstitution, countryOfHostInstitution);
        this.setCountryOfWorkPlacement(countryOfWorkPlacement);
        this.setLengthStudyPeriod(lengthStudyPeriod);
        this.setLengthWorkPlacement(lengthWorkPlacement);
        this.setStudyStartDate(studyStartDate);
        this.setPlacementStartDate(placementStartDate);
        this.setTaughtHostLang(taughtHostLang);
        this.setLanguageTaught(languageTaught);
        this.setStudyGrant(studyGrant);
        this.setPlacementGrant(placementGrant);
    }
    
    public String getHomeInstitution() {
    	return this.homeInstitution;
    }
    
    public void setHomeInstitution(String homeInstitution) {
    	this.homeInstitution = homeInstitution;
    }
    
    public String getCountryOfHomeInstitution() {
    	return this.countryOfHomeInstitution;
    }
    
    public void setCountryOfHomeInstitution(String countryOfHomeInstitution) {
    	this.countryOfHomeInstitution = countryOfHomeInstitution;
    }
    
    public Integer getAge() {
    	return this.age;
    }
    
    public void setAge(Integer age) {
    	this.age = age;
    }
    
    public Gender getGender() {
    	return this.gender;
    }
    
    public void setGender(Gender gender) {
    	this.gender = gender;
    }
    
    public StudyLevel getStudyLevel() {
    	return this.studyLevel;
    }
    
    public void setStudyLevel(StudyLevel studyLevel) {
    	this.studyLevel = studyLevel;
    }
    
    public MobilityType getMobilityType() {
    	return this.mobilityType;
    }
    
    public void setMobilityType(MobilityType mobilityType) {
    	this.mobilityType = mobilityType;
    }
    
    public String getHostInstitution() {
    	return this.hostInstitution;
    }
    
    public void setHostInstitution(String hostInstitution) {
    	this.hostInstitution = hostInstitution;
    }
    
    public String getCountryOfHostInstitution() {
    	return this.countryOfHostInstitution;
    }
    
    public void setCountryOfHostInstitution(String countryOfHostInstitution) {
    	this.countryOfHostInstitution = countryOfHostInstitution;
    }
    
    public String getCountryOfWorkPlacement() {
    	return this.countryOfWorkPlacement;
    }
    
    public void setCountryOfWorkPlacement(String countryOfWorkPlacement) {
    	this.countryOfWorkPlacement = countryOfWorkPlacement;
    }
    
    public Float getLengthStudyPeriod() {
    	return this.lengthStudyPeriod;
    }
    
    public void setLengthStudyPeriod(Float lengthStudyPeriod) {
    	this.lengthStudyPeriod = lengthStudyPeriod;
    }
    
    public Float getLengthWorkPlacement() {
    	return this.lengthWorkPlacement;
    }
    
    public void setLengthWorkPlacement(Float lengthWorkPlacement) {
    	this.lengthWorkPlacement = lengthWorkPlacement;
    }
    
    public LocalDate getStudyStartDate() {
    	return this.studyStartDate;
    }
    
    public void setStudyStartDate(LocalDate studyStartDate) {
    	this.studyStartDate = studyStartDate;
    }
    
    public LocalDate getPlacementStartDate() {
    	return this.placementStartDate;
    }
    
    public void setPlacementStartDate(LocalDate placementStartDate) {
    	this.placementStartDate = placementStartDate;
    }
    
    public Boolean getTaughtHostLang() {
    	return this.taughtHostLang;
    }
    
    public void setTaughtHostLang(Boolean taughtHostLang) {
    	this.taughtHostLang = taughtHostLang;
    }
    
    public String getLanguageTaught() {
    	return this.languageTaught;
    }
    
    public void setLanguageTaught(String languageTaught) {
    	this.languageTaught = languageTaught;
    }
    
    public Float getStudyGrant() {
    	return this.studyGrant;
    }
    
    public void setStudyGrant(Float studyGrant) {
    	this.studyGrant = studyGrant;
    }
    
    public Float getPlacementGrant() {
    	return this.placementGrant;
    }
    
    public void setPlacementGrant(Float placementGrant) {
    	this.placementGrant = placementGrant;
    }

    /**
     * Returns a string with all the Student properties joined by semicolons.
     * It applies default values to some attributes if they are missing.
     * The format of the string returned is analogous to the one admitted by the
     * @code{fromString(String)} method.
     */
    @Override
    public String toString() {
        StringBuilder str = new StringBuilder();
        
        str.append(this.getHomeInstitution()).append(SPLIT);
        str.append(this.getCountryOfHomeInstitution()).append(SPLIT);
        str.append(this.getAge()).append(SPLIT);
        str.append(this.getGender()).append(SPLIT);
        str.append(this.getStudyLevel()).append(SPLIT);
        str.append(this.getMobilityType()).append(SPLIT);
        str.append(this.getHostInstitution());
        
        String countryOfHostInstitution = this.getCountryOfHostInstitution();
        String countryOfWorkPlacement = this.getCountryOfWorkPlacement();
        Float lengthStudyPeriod = this.getLengthStudyPeriod();
        Float lengthWorkPlacement = this.getLengthWorkPlacement();
        LocalDate studyStartDate = this.getStudyStartDate();
        LocalDate placementStartDate = this.getPlacementStartDate();
        Boolean taughtHostLang = this.getTaughtHostLang();
        String languageTaught = this.getLanguageTaught();
        Float studyGrant = this.getStudyGrant();
        Float placementGrant = this.getPlacementGrant();

        str.append(SPLIT).append(countryOfHostInstitution != null ? countryOfHostInstitution : DEFAULT_STR);
        str.append(SPLIT).append(countryOfWorkPlacement != null ? countryOfWorkPlacement : DEFAULT_STR);
        str.append(SPLIT).append(String.valueOf(lengthStudyPeriod != null ? lengthStudyPeriod : DEFAULT_INT));
        str.append(SPLIT).append(String.valueOf(lengthWorkPlacement != null ? lengthWorkPlacement : DEFAULT_INT));
        str.append(SPLIT).append(studyStartDate != null ? studyStartDate.toString() : DEFAULT_STR);
        str.append(SPLIT).append(placementStartDate != null ? placementStartDate.toString() : DEFAULT_STR);
        str.append(SPLIT).append(String.valueOf(taughtHostLang != null ? taughtHostLang : DEFAULT_BOOL));
        str.append(SPLIT).append(languageTaught != null ? languageTaught : DEFAULT_STR);
        str.append(SPLIT).append(String.valueOf(studyGrant != null ? studyGrant : DEFAULT_INT));
        str.append(SPLIT).append(String.valueOf(placementGrant != null ? placementGrant : DEFAULT_INT));
        
        return str.toString();
    }
    
    /**
     * Convenience method for building a Student instance from a semicolon separated string
     * containing the object's properties.
     */
    public static Student fromString(String stringifiedStudent) throws StudentParseException {
    	Student student;
    	int currentField = 0;
    	String[] studentFields = stringifiedStudent.split(SPLIT);
    	
    	// Neither basic nor full student
    	if (studentFields.length < NUM_ATTRIBUTES_BASIC - 1
    			|| studentFields.length > NUM_ATTRIBUTES_FULL) {
    		System.err.println(String.format("%s: %s",
    				StudentParseException.class.getName(),
    				StudentParseException.forInputString(stringifiedStudent).getMessage()));
    		return new Student();
    	}
    	
    	try {
	    	String homeInstitution = studentFields[currentField];
	    	String countryOfHomeInstitution = studentFields[++currentField];
	    	Integer age = studentFields[++currentField].length() > 0 ? Integer.valueOf(studentFields[currentField]) : null;
	    	Gender gender = Gender.fromValue(studentFields[++currentField].charAt(0));
	    	StudyLevel studyLevel = StudyLevel.fromValue(studentFields[++currentField].charAt(0));
	    	MobilityType mobilityType = MobilityType.fromValue(studentFields[++currentField].charAt(0));
	    	String hostInstitution = studentFields[++currentField];
	    	String countryOfHostInstitution = studentFields.length >= NUM_ATTRIBUTES_BASIC ? studentFields[++currentField] : EMPTY_STRING;
	    	
	    	if (studentFields.length <= NUM_ATTRIBUTES_BASIC) {
	    		student = new Student(homeInstitution, countryOfHomeInstitution,
	    				age, gender, studyLevel, mobilityType, hostInstitution, countryOfHostInstitution);
	    	} else {
	    		String countryOfWorkPlacement = studentFields[++currentField];
	    		Float lengthStudyPeriod = studentFields[++currentField].length() > 0 ? parseFloat(studentFields[currentField]) : null;
	    		Float lengthWorkPlacement = studentFields[++currentField].length() > 0 ? parseFloat(studentFields[currentField]) : null;
	    		LocalDate studyStartDate = dateFromString(studentFields[++currentField]);
	    		LocalDate placementStartDate = dateFromString(studentFields[++currentField]);
	    		Boolean taughtHostLang = boolFromChar(studentFields[++currentField]);
	    		
	    		String languageTaught = studentFields[++currentField];
	    		Float studyGrant = studentFields.length > ++currentField && studentFields[currentField].length() > 0 ?
	    				parseFloat(studentFields[currentField]) : null;
	    		Float placementGrant = studentFields.length > ++currentField && studentFields[currentField].length() > 0 ?
	    				parseFloat(studentFields[currentField]) : null;
	    		
	    		student = new Student(homeInstitution, countryOfHomeInstitution,
	    				age, gender, studyLevel, mobilityType, hostInstitution,
	    				countryOfHostInstitution, countryOfWorkPlacement,
	    				lengthStudyPeriod, lengthWorkPlacement,
	    				studyStartDate, placementStartDate, taughtHostLang,
	    				languageTaught, studyGrant, placementGrant);
	    	}
    	} catch (Exception e) {
    		System.err.println(stringifiedStudent);
    		System.err.println(String.format("%s: %s", e.getClass().getName(), e.getMessage()));
    		return new Student();
    	}
    	
    	return student;
    }
    
    private static final char MALE_VALUE = 'M';
    private static final char FEMALE_VALUE = 'F';
    
    public static enum Gender {
    	
		MALE("MALE", MALE_VALUE),
		FEMALE("FEMALE", FEMALE_VALUE);
		
		private char value;
		private String name;
		
		Gender(String name, char value) {
			this.name = name;
			this.value = value;
		}
		
		public static Gender fromValue(char value) {
			Gender gender = null;
			switch (value) {
			case MALE_VALUE:
				gender = Gender.MALE;
				break;
			case FEMALE_VALUE:
				gender = Gender.FEMALE;
				break;
			}
			return gender;
		}
		
		public char getValue() {
			return this.value;
		}
		
		@Override
		public String toString() {
			return this.name;
		}
    }
    
    private static final char FIRST_VALUE = '1';
    private static final char SECOND_VALUE = '2';
    private static final char SUPERIOR_VALUE = 'S';
    
    public static enum StudyLevel {
    	
    	FIRST("FIRST", FIRST_VALUE),
    	SECOND("SECOND", SECOND_VALUE),
    	SUPERIOR("SUPERIOR", SUPERIOR_VALUE);
    	
    	private char value;
    	private String name;
    	
    	StudyLevel(String name, char value) {
    		this.name = name;
    		this.value = value;
    	}
    	
    	public static StudyLevel fromValue(char value) {
    		StudyLevel sl = null;
    		switch (value) {
    		case FIRST_VALUE:
    			sl = StudyLevel.FIRST;
    			break;
    		case SECOND_VALUE:
    			sl = StudyLevel.SECOND;
    			break;
    		case SUPERIOR_VALUE:
    			sl = StudyLevel.SUPERIOR;
    			break;
    		}
    		return sl;
    	}
    	
    	public char getValue() {
    		return this.value;
    	}
    	
    	@Override
    	public String toString() {
    		return this.name;
    	}
    }
    
    private static final char STUDY_VALUE = 'S';
    private static final char PLACEMENT_VALUE = 'P';
    private static final char OTHER_VALUE = 'C';
    
    public static enum MobilityType {
    	
    	STUDY("STUDY", STUDY_VALUE),
    	PLACEMENT("PLACEMENT", PLACEMENT_VALUE),
    	OTHER("OTHER", OTHER_VALUE);
    	
    	private char value;
    	private String name;
    	
    	MobilityType(String name, char value) {
    		this.name = name;
    		this.value = value;
    	}
    	
    	public static MobilityType fromValue(char value) {
    		MobilityType mt = null;
    		switch (value) {
    		case STUDY_VALUE:
    			mt = MobilityType.STUDY;
    			break;
    		case PLACEMENT_VALUE:
    			mt = MobilityType.PLACEMENT;
    			break;
    		case OTHER_VALUE:
    			mt = MobilityType.OTHER;
    			break;
    		}
    		return mt;
    	}
    	
    	public char getValue() {
    		return this.value;
    	}
    	
    	@Override
    	public String toString() {
    		return this.name;
    	}
    }
    
    private static Boolean boolFromChar(String input) {
    	final char YES = 'Y';
    	final char NO = 'N';
    	
    	boolean output = DEFAULT_BOOL;
    	if (input.length() > 0) {
    		switch (input.charAt(0)) {
    		case YES:
    			output = true;
    			break;
    		case NO:
    			output = false;
    			break;
    		}
		}
    	return output;
    }
    
    private static LocalDate dateFromString(String input) throws DateTimeParseException {
    	final char split = input.indexOf('-') > 0 ? '-' : '/';
    	LocalDate output = null;
    	String pattern = null;
    	
    	switch (input.length()) {
    	case 0:
    		break;
    	case 7:
    		if (!input.endsWith("0000")) {
	    		input = "1" + split + input;
	    		pattern = "d" + split + "MM" + split + "yyyy";
    		}
    		break;
    	case 20:
    		pattern = "dd-MMM-yyyy HH.mm.ss";
    		break;
    	}
    	
    	if (pattern != null) {
    		DateTimeFormatter formatter = new DateTimeFormatterBuilder().parseCaseInsensitive()
    				.appendPattern(pattern).toFormatter(Locale.ENGLISH);
    		output = LocalDate.parse(input, formatter);
    	}
    	return output;
    }
    
    private static Float parseFloat(String input) throws NumberFormatException {
    	Float output = null;
    	if (input.length() > 0) {
    		input = input.replace(',', '.');
    		output = Float.valueOf(input);
    	}
    	return output;
    }
    
    public static class StudentParseException extends IllegalArgumentException {
    	
		private static final long serialVersionUID = 4604040086577779690L;

		public StudentParseException() {
    		super();
    	}
    	
    	public StudentParseException(String s) {
    		super(s);
    	}
    	
    	static StudentParseException forInputString(String s) {
    		return new StudentParseException("For input string: \"" + s + "\"");
    	}
    }
}

File: pandas/domain/student.py
"""pandas - domain/student.py

Student Mobility

Javier Fernández-Bravo Peñuela
Universitat Politècnica de València
"""

import locale
from datetime import date, datetime
from enum import Enum

#locale.setlocale(locale.LC_NUMERIC, "en_us")
#locale.setlocale(locale.LC_TIME, "en_us")

class Student:
    def __init__(self, home_institution, country_of_home_institution, age,
                gender, study_level, mobility_type, host_institution, country_of_host_institution,
                country_of_work_placement, length_study_period=None, length_work_placement=None,
                study_start_date=None, placement_start_date=None, taught_host_lang=None,
                language_taught=None, study_grant=None, placement_grant=None):
        self.set_home_institution(home_institution)
        self.set_country_of_home_institution(country_of_home_institution)
        self.set_age(age)
        self.set_gender(gender)
        self.set_study_level(study_level)
        self.set_mobility_type(mobility_type)
        self.set_host_institution(host_institution)
        self.set_country_of_host_institution(country_of_host_institution)
        self.set_country_of_work_placement(country_of_work_placement)

        self.set_length_study_period(length_study_period)
        self.set_length_work_placement(length_work_placement)
        self.set_study_start_date(study_start_date)
        self.set_placement_start_date(placement_start_date)
        self.set_taught_host_lang(taught_host_lang)
        self.set_language_taught(language_taught)
        self.set_study_grant(study_grant)
        self.set_placement_grant(placement_grant)
    
    def get_home_institution(self):
        return self.home_institution
    
    def set_home_institution(self, home_institution):
        self.home_institution = string_factory(home_institution)
    
    def get_country_of_home_institution(self):
        return self.country_of_home_institution
    
    def set_country_of_home_institution(self, country_of_home_institution):
        self.country_of_home_institution = string_factory(country_of_home_institution)
    
    def get_age(self):
        return self.age
    
    def set_age(self, age):
        self.age = integer_factory(age)
    
    def get_gender(self):
        return self.gender
    
    def set_gender(self, gender):
        self.gender = gender_factory(gender)
    
    def get_study_level(self):
        return self.study_level
    
    def set_study_level(self, study_level):
        self.study_level = study_level_factory(study_level)
    
    def get_mobility_type(self):
        return self.mobility_type
    
    def set_mobility_type(self, mobility_type):
        self.mobility_type = mobility_type_factory(mobility_type)
    
    def get_host_institution(self, host_institution):
        return self.host_institution

    def set_host_institution(self, host_institution):
        self.host_institution = string_factory(host_institution)
    
    def get_country_of_host_institution(self):
        return self.country_of_host_institution
    
    def set_country_of_host_institution(self, country_of_host_institution):
        self.country_of_host_institution = string_factory(country_of_host_institution)
    
    def get_country_of_work_placement(self):
        return self.country_of_work_placement
    
    def set_country_of_work_placement(self, country_of_work_placement):
        self.country_of_work_placement = string_factory(country_of_work_placement)
    
    def get_length_study_period(self):
        return self.length_study_period
    
    def set_length_study_period(self, length_study_period):
        self.length_study_period = float_factory(length_study_period)
    
    def get_length_work_placement(self):
        return self.length_work_placement
    
    def set_length_work_placement(self, length_work_placement):
        self.length_work_placement = float_factory(length_work_placement)
    
    def get_study_start_date(self):
        return self.study_start_date
    
    def set_study_start_date(self, study_start_date):
        self.study_start_date = date_factory(study_start_date)
    
    def get_placement_start_date(self):
        return self.study_start_date
    
    def set_placement_start_date(self, placement_start_date):
        self.placement_start_date = date_factory(placement_start_date)
    
    def get_taught_host_lang(self):
        return self.taught_host_lang
    
    def set_taught_host_lang(self, taught_host_lang):
        self.taught_host_lang = boolean_factory(taught_host_lang)
    
    def get_language_taught(self):
        return self.language_taught
    
    def set_language_taught(self, language_taught):
        self.language_taught = string_factory(language_taught)
    
    def get_study_grant(self):
        return self.study_grant
    
    def set_study_grant(self, study_grant):
        self.study_grant = float_factory(study_grant)
    
    def get_placement_grant(self):
        return self.placement_grant
    
    def set_placement_grant(self, placement_grant):
        self.placement_grant = float_factory(placement_grant)
    
    def to_string(self):
        def default_string(string): return string if string is not None else ''
        def default_integer(integer): return integer if integer is not None else 0.
        def default_date(date_value): return date_value.isoformat() if date_value is not None else ''
        def default_boolean(boolean): return boolean if boolean is not None else False

        return ';'.join([
            self.home_institution,
            self.country_of_home_institution,
            self.age,
            self.gender,
            self.study_level,
            self.mobility_type,
            self.host_institution,
            default_string(self.country_of_host_institution),
            default_string(self.country_of_work_placement),
            default_integer(self.length_study_period),
            default_integer(self.length_work_placement),
            default_date(self.study_start_date),
            default_date(self.placement_start_date),
            default_boolean(self.taught_host_lang),
            default_string(self.language_taught),
            default_integer(self.study_grant),
            default_integer(self.placement_grant)
        ])
    
    @staticmethod
    def from_string(stringified_student):
        num_attributes_basic = 9
        num_attributes_full = 17
        student_fields = stringified_student.split(';')
        if len(student_fields) < num_attributes_basic or len(student_fields) > num_attributes_full:
            return None
        
        converters = Student.get_converters()

        (home_institution,
        country_of_home_institution,
        age,
        gender,
        study_level,
        mobility_type,
        host_institution,
        country_of_host_institution,
        country_of_work_placement) = [converters[n](student_fields[n]) for n in range(num_attributes_basic)]

        if len(student_fields) < num_attributes_full:
            student = Student(home_institution, country_of_home_institution, age, gender,
                            study_level, mobility_type, host_institution,
                            country_of_host_institution, country_of_work_placement)
        else:
            (length_study_period,
            length_work_placement,
            study_start_date,
            placement_start_date,
            taught_host_lang,
            language_taught,
            study_grant,
            placement_grant) = [converters[n](student_fields[n]) for n in range(num_attributes_basic, num_attributes_full)]
            student = Student(home_institution, country_of_home_institution, age, gender,
                            study_level, mobility_type, host_institution,
                            country_of_host_institution, country_of_work_placement,
                            length_study_period, length_work_placement,
                            study_start_date, placement_start_date, taught_host_lang,
                            language_taught, study_grant, placement_grant)
        return student
    
    @staticmethod
    def get_converters():
        return [
            string_factory,         # 0:  home_institution
            string_factory,         # 1:  country_of_home_institution
            integer_factory,        # 2:  age
            gender_factory,         # 3:  gender
            study_level_factory,    # 4:  study_level
            mobility_type_factory,  # 5:  mobility_type
            string_factory,         # 6:  host_institution
            string_factory,         # 7:  country_of_host_institution
            string_factory,         # 8:  country_of_work_placement
            float_factory,          # 9:  length_study_period
            float_factory,          # 10: length_work_placement
            date_factory,           # 11: study_start_date
            date_factory,           # 12: placement_start_date
            boolean_factory,        # 13: taught_host_lang
            string_factory,         # 14: language_taught
            float_factory,          # 15: study_grant
            float_factory           # 16: placement_grant
        ]

class Gender(Enum):
    MALE = 'M'
    FEMALE = 'F'

class StudyLevel(Enum):
    FIRST = '1'
    SECOND = '2'
    SUPERIOR = 'S'

class MobilityType(Enum):
    STUDY = 'S'
    PLACEMENT = 'P'
    OTHER = 'C'

def string_factory(string):
    return str(string) if not isNone(string) else ''

def integer_factory(integer):
    return int(integer) if not isNone(integer) else 0

def float_factory(float_number):
    return parse_float(float_number) if not isNone(float_number) else 0.

def boolean_factory(boolean):
    return bool_from_char(boolean)

def date_factory(date_string):
    return date_from_string(date_string)

def gender_factory(gender):
    return gender_from_value(gender)

def study_level_factory(study_level):
    return study_level_from_value(study_level)

def mobility_type_factory(mobility_type):
    return mobility_type_from_value(mobility_type)

def bool_from_char(input):
    YES = 'Y'
    NO = 'N'
    
    output = False
    if isinstance(input, str) and len(input) > 0:
        if input[0] == YES:
            output = True
        elif input[0] == NO:
            output = False
    return output

def date_from_string(input):
    if isinstance(input, date):
        return input
    elif isNone(input):
        return None
    
    split = '/' if '/' in input else '-'
    pattern = None
    output = None
    if len(input) == 7:
        if not input.endswith("0000"):
            input = '01' + split + input
            pattern = '%d' + split + '%m' + split + '%Y'
    elif len(input) == 20:
        pattern = '%d' + split + '%b' + split + '%Y %H.%M.%S'
    
    if pattern is not None:
        output = datetime.strptime(input.lower(), pattern).date()
    return output

def parse_float(input):
    if isinstance(input, str):
        input = input.replace(',', '.')
    output = float(input)
    return output

def gender_from_value(gender_value):
    gender_output = None
    if isinstance(gender_value, Gender):
        gender_output = Gender
    elif isinstance(gender_value, str) and len(gender_value) == 1:
        if gender_value == Gender.MALE.value:
            gender_output = Gender.MALE
        elif gender_value == Gender.FEMALE.value:
            gender_output = Gender.FEMALE
    return gender_output

def study_level_from_value(study_level_value):
    study_level_output = None
    if isinstance(study_level_value, StudyLevel):
        study_level_output = study_level_value
    elif not isNone(study_level_value):
        study_level_value = str(study_level_value)
        if study_level_value == StudyLevel.FIRST.value:
            study_level_output = StudyLevel.FIRST
        elif study_level_value == StudyLevel.SECOND.value:
            study_level_output = StudyLevel.SECOND
        elif study_level_value == StudyLevel.SUPERIOR.value:
            study_level_output = StudyLevel.SUPERIOR
    return study_level_output

def mobility_type_from_value(mobility_type_value):
    mobility_type_output = None
    if isinstance(mobility_type_value, MobilityType):
        mobility_type_output = mobility_type_value
    elif not isNone(mobility_type_value):
        mobility_type_value = str(mobility_type_value)
        if mobility_type_value == MobilityType.STUDY.value:
            mobility_type_output = MobilityType.STUDY
        elif mobility_type_value == MobilityType.PLACEMENT.value:
            mobility_type_output = MobilityType.PLACEMENT
        elif mobility_type_value == MobilityType.OTHER.value:
            mobility_type_output = MobilityType.OTHER
    return mobility_type_output

def isNone(arg):
    return arg is None or arg == ''
